### PR TITLE
Add osm notes importing

### PR DESCRIPTION
**Pull request by nnngrach**
https://github.com/osmandapp/OsmAnd-Issues/issues/275

### CODE REVIEW

File: OsmAnd.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -5102,8 +5102,8 @@
 		DA6D07A025518D90004FE017 /* OAInfoBottomView.xib in Resources */ = {isa = PBXBuildFile; fileRef = DA6D079E25518D90004FE017 /* OAInfoBottomView.xib */; };
 		DA6D07A525543735004FE017 /* OAClearPointsCommand.mm in Sources */ = {isa = PBXBuildFile; fileRef = DA6D07A425543735004FE017 /* OAClearPointsCommand.mm */; };
 		DA6D07A625543735004FE017 /* OAClearPointsCommand.mm in Sources */ = {isa = PBXBuildFile; fileRef = DA6D07A425543735004FE017 /* OAClearPointsCommand.mm */; };
-		DA6D51E321F328AA00EC84E2 /* OAOsmEditsDBHelper.m in Sources */ = {isa = PBXBuildFile; fileRef = DA6D51E221F328AA00EC84E2 /* OAOsmEditsDBHelper.m */; };
-		DA6D51E421F328AA00EC84E2 /* OAOsmEditsDBHelper.m in Sources */ = {isa = PBXBuildFile; fileRef = DA6D51E221F328AA00EC84E2 /* OAOsmEditsDBHelper.m */; };
+		DA6D51E321F328AA00EC84E2 /* OpenstreetmapsDbHelper.m in Sources */ = {isa = PBXBuildFile; fileRef = DA6D51E221F328AA00EC84E2 /* OpenstreetmapsDbHelper.m */; };
```
Comment (PaulStets) [inline OsmAnd.xcodeproj/project.pbxproj:5105]: All classes should have ```OA``` prefix! Furthermore, these renamings are unnecessary as the class names were created by me to improve code readability and fix grammatical errors introduced in Android.

File: Sources/Controllers/2.0/BottomSheet/OsmEditingBottomSheet/OAOsmNoteBottomSheetViewController.mm [hunk 2]
```diff
@@ -212,8 +212,8 @@ -(void) doneButtonPressed
 - (void) saveNote
 {
     dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-        id<OAOsmBugsUtilsProtocol> util = [_plugin getLocalOsmNotesUtil];
-        OAOsmNotePoint *p = _bugPoints.firstObject;
+        id<OAOsmBugsUtilsProtocol> util = [_plugin getOsmNotesLocalUtil];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/BottomSheet/OsmEditingBottomSheet/OAOsmNoteBottomSheetViewController.mm:215]: Again, pointless renaming of the proper grammar. This was not in the task.

File: Sources/Controllers/2.0/Settings/OAImportCompleteViewController.mm [hunk 3]
```diff
@@ -341,6 +355,13 @@ - (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath
         UIViewController* resourcesViewController = [[UIStoryboard storyboardWithName:@"Resources" bundle:nil] instantiateInitialViewController];
         [rootController.navigationController pushViewController:resourcesViewController animated:YES];
     }
+    else if (dataType == EOAImportDataTypeOsmNotes)
+    {
+
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/OAImportCompleteViewController.mm:360]: Remove extra line

File: Sources/Controllers/2.0/Settings/OAImportCompleteViewController.mm [hunk 4]
```diff
@@ -341,6 +355,13 @@ - (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath
         UIViewController* resourcesViewController = [[UIStoryboard storyboardWithName:@"Resources" bundle:nil] instantiateInitialViewController];
         [rootController.navigationController pushViewController:resourcesViewController animated:YES];
     }
+    else if (dataType == EOAImportDataTypeOsmNotes)
+    {
+        
+        UITabBarController* myPlacesViewController = [[UIStoryboard storyboardWithName:@"MyPlaces" bundle:nil] instantiateInitialViewController];
+        [myPlacesViewController setSelectedIndex:2];
```
Comment (PaulStets) [inline Sources/Controllers/2.0/Settings/OAImportCompleteViewController.mm:362]: Check if OSM Editing plugin is enabled before opening the screen. If the plugin is disabled this will cause a crash.

File: Sources/Helpers/OASettingsHelper.mm [hunk 5]
```diff
@@ -1148,7 +1160,7 @@ - (OASettingsItemWriter *)getWriter
 @implementation OAPluginSettingsItem
 {
     OAPlugin *_plugin;
-    NSArray<OASettingsItem *> *_pluginDependentItems;
+    NSMutableArray<OASettingsItem *> *_pluginDependentItems;
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:1163]: Why did you change OAPluginSettingsItem in this task?

File: Sources/Helpers/OASettingsHelper.mm [hunk 6]
```diff
@@ -2688,3 +2705,151 @@ - (void) writeItemsToJson:(id)json error:(NSError * _Nullable __autoreleasing *)
...
+
+@property (nonatomic) NSMutableArray<OAOsmNotesPoint *> *items;
+@property (nonatomic) NSMutableArray<OAOsmNotesPoint *> *appliedItems;
+@property (nonatomic) NSMutableArray<OAOsmNotesPoint *> *duplicateItems;
+@property (nonatomic) NSMutableArray<OAOsmNotesPoint *> *existingItems;
+
+@end
+
+@implementation OAOsmNotesSettingsItem
+{
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:2721]: No need for braces if the implementation does not have private fields.

File: Sources/Helpers/OASettingsHelper.mm [hunk 7]
```diff
@@ -2688,3 +2705,151 @@ - (void) writeItemsToJson:(id)json error:(NSError * _Nullable __autoreleasing *)
...
+}
+
+@dynamic items, appliedItems, duplicateItems, existingItems;
+
+NSString *ID_KEY = @"id";
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsHelper.mm:2726]: Make the variables ```static```

File: Sources/Plugins/OsmEditing/OAOsmEditingPlugin.mm [hunk 8]
```diff
@@ -110,41 +121,46 @@ - (void) updateLayers
...
 
--(id<OAOsmBugsUtilsProtocol>)getLocalOsmNotesUtil
+-(id<OAOsmBugsUtilsProtocol>)getOsmNotesLocalUtil
 {
-    return _localBugsUtil;
+    return _localNotesUtil;
 }
 
--(id<OAOsmBugsUtilsProtocol>)getRemoteOsmNotesUtil
+- (OAOsmBugsDBHelper *) getDBBug
```
Comment (PaulStets) [inline Sources/Plugins/OsmEditing/OAOsmEditingPlugin.mm:149]: There is absolutely no need for this method in iOS. Objective C's singletons are easy to use without the need to call such methods (unlike Java).

File: Sources/Controllers/2.0/Settings/OAImportCompleteViewController.mm (hunk 4 above)
Comment (nnngrach) [reply Sources/Controllers/2.0/Settings/OAImportCompleteViewController.mm:362]: It's ok. If osm plugin is disabled then `[myPlacesViewController setSelectedIndex:2]` redirects to screen with index 0. There's no crash.

File: OsmAnd.xcodeproj/project.pbxproj (hunk 1 above)
Comment (nnngrach) [reply OsmAnd.xcodeproj/project.pbxproj:5105]: renamed

File: Sources/Controllers/2.0/Settings/OAImportCompleteViewController.mm (hunk 4 above)
Comment (nnngrach) [reply Sources/Controllers/2.0/Settings/OAImportCompleteViewController.mm]: Check added&

File: Sources/Helpers/OASettingsImporter.mm [hunk 9]
```diff
@@ -147,7 +147,6 @@ - (void) importItems:(NSString *)file items:(NSArray<OASettingsItem *> *)items
             NSString *itemsJson = [NSString stringWithContentsOfFile:tmpFileName encoding:NSUTF8StringEncoding error:nil];
             OASettingsItemsFactory *factory = [[OASettingsItemsFactory alloc] initWithJSON:itemsJson];
             [items addObjectsFromArray:factory.getItems];
-            [fileManager removeItemAtPath:_tmpFilesDir error:nil];
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsImporter.mm:150]: Check if items are deleted after the import is complete

File: Sources/Helpers/OASettingsImporter.mm [hunk 10]
```diff
@@ -187,74 +186,51 @@ - (void) collectItems:(NSString *)jsonStr
...
     {
         NSLog(@"Error: unsupported version");
         return;
     }
-    
-    NSArray* itemsJson = json[@"items"];
-//    NSMutableDictionary *pluginItems = [NSMutableDictionary new];
-    
-    for (NSDictionary* item in itemsJson)
+    NSMutableDictionary *pluginItems = [NSMutableDictionary new];
```
Comment (PaulStets) [inline Sources/Helpers/OASettingsImporter.mm:196]: Plugins have not been implemented yet. Revert this code to its original state
